File: ios/Omniverse/Storage/Stores.swift
import Foundation
import Security

// MARK: - Keychain-backed credentials store (mirrors CredentialsStore)

struct CredentialsStore {
    private let service = "com.aryaroop.omniverse.credentials"

    private enum K {
        static let tmdbToken = "tmdb_token"
        static let tvdbKey = "tvdb_api_key"
        static let tvdbPin = "tvdb_pin"
        static let traktClientId = "trakt_client_id"
        static let traktClientSecret = "trakt_client_secret"
        static let traktAccessToken = "trakt_access_token"
        static let traktRefreshToken = "trakt_refresh_token"
        static let traktTokenExpiresAt = "trakt_token_expires_at"
        static let traktUsername = "trakt_username"
        static let pixeldrainApiKey = "pixeldrain_api_key"
        static let anilistAccessToken = "anilist_access_token"
    }

    func load() -> ApiCredentials {
        var c = ApiCredentials()
        c.tmdbToken = read(K.tmdbToken) ?? ""
        c.tvdbApiKey = read(K.tvdbKey) ?? ""
        c.tvdbPin = read(K.tvdbPin) ?? ""
        c.traktClientId = read(K.traktClientId) ?? ""
        c.traktClientSecret = read(K.traktClientSecret) ?? ""
        c.traktAccessToken = read(K.traktAccessToken) ?? ""
        c.traktRefreshToken = read(K.traktRefreshToken) ?? ""
        c.traktTokenExpiresAt = Int(read(K.traktTokenExpiresAt) ?? "") ?? 0
        c.traktUsername = read(K.traktUsername) ?? ""
        c.pixeldrainApiKey = read(K.pixeldrainApiKey) ?? ""
        c.anilistAccessToken = read(K.anilistAccessToken) ?? ""
        return c
    }

    func save(_ c: ApiCredentials) {
        write(K.tmdbToken, c.tmdbToken.trimmed)
        write(K.tvdbKey, c.tvdbApiKey.trimmed)
        write(K.tvdbPin, c.tvdbPin.trimmed)
        write(K.traktClientId, c.traktClientId.trimmed)
        write(K.traktClientSecret, c.traktClientSecret.trimmed)
        write(K.traktAccessToken, c.traktAccessToken.trimmed)
        write(K.traktRefreshToken, c.traktRefreshToken.trimmed)
        write(K.traktTokenExpiresAt, String(c.traktTokenExpiresAt))
        write(K.traktUsername, c.traktUsername.trimmed)
        write(K.pixeldrainApiKey, c.pixeldrainApiKey.trimmed)
        write(K.anilistAccessToken, c.anilistAccessToken.trimmed)
    }

    private func read(_ key: String) -> String? {
        let q: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var out: AnyObject?
        guard SecItemCopyMatching(q as CFDictionary, &out) == errSecSuccess,
              let data = out as? Data else { return nil }
        return String(data: data, encoding: .utf8)
    }

    private func write(_ key: String, _ value: String) {
        let base: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
        ]
        SecItemDelete(base as CFDictionary)
        var add = base
        add[kSecValueData as String] = Data(value.utf8)
        add[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlock
        SecItemAdd(add as CFDictionary, nil)
    }
}

// MARK: - UserDefaults-backed settings/cache store (mirrors UserSettingsStore)

struct UserSettingsStore {
    private let d = UserDefaults.standard
    private enum K {
        static let settings = "settings"
        static let liveTvSources = "live_tv_sources"
        static let cachedCategories = "cached_categories"
        static let cachedLiveTv = "cached_live_tv"
        static let watchlist = "watchlist"
        static let watchHistory = "watch_history_v1"
    }

    func loadSettings() -> UserSettings {
        guard let data = d.data(forKey: K.settings),
              let s = try? JSONDecoder().decode(UserSettings.self, from: data) else {
            return UserSettings()
        }
        return s
    }
    func saveSettings(_ s: UserSettings) { d.set(try? JSONEncoder().encode(s), forKey: K.settings) }

    func loadLiveTvSources() -> [LiveTvSource] { decodeArray(K.liveTvSources) }
    func saveLiveTvSources(_ v: [LiveTvSource]) { encodeArray(v, K.liveTvSources) }

    func loadCachedCategories() -> [MediaCategory] { decodeArray(K.cachedCategories) }
    func saveCachedCategories(_ v: [MediaCategory]) { encodeArray(v, K.cachedCategories) }

    func loadCachedLiveTv() -> [LiveTvEntry] { decodeArray(K.cachedLiveTv) }
    func saveCachedLiveTv(_ v: [LiveTvEntry]) { encodeArray(v, K.cachedLiveTv) }

    func loadWatchlist() -> Set<String> { Set(d.stringArray(forKey: K.watchlist) ?? []) }
    func saveWatchlist(_ v: Set<String>) { d.set(v.sorted(), forKey: K.watchlist) }

    func loadWatchHistory() -> [WatchProgress] { decodeArray(K.watchHistory) }
    func saveWatchHistory(_ v: [WatchProgress]) { encodeArray(v, K.watchHistory) }

    func lastRefreshedTime() -> Int { d.integer(forKey: "last_refreshed_time") }
    func setLastRefreshedTime(_ ms: Int) { d.set(ms, forKey: "last_refreshed_time") }

    private func decodeArray<T: Decodable>(_ key: String) -> [T] {
        guard let data = d.data(forKey: key),
              let v = try? JSONDecoder().decode([T].self, from: data) else { return [] }
        return v
    }
    private func encodeArray<T: Encodable>(_ v: [T], _ key: String) {
        d.set(try? JSONEncoder().encode(v), forKey: key)
    }
}

File: android/app/src/main/java/com/finix/omniverse/Stores.kt
package com.finix.omniverse

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.json.Json

private val json = Json {
    ignoreUnknownKeys = true
    encodeDefaults = true
}

// MARK: - EncryptedSharedPreferences-backed credentials store
// Mirrors the iOS CredentialsStore (Keychain). Same key names as the Dart app.

class CredentialsStore(context: Context) {

    private object K {
        const val tmdbToken = "tmdb_token"
        const val tvdbKey = "tvdb_api_key"
        const val tvdbPin = "tvdb_pin"
        const val traktClientId = "trakt_client_id"
        const val traktClientSecret = "trakt_client_secret"
        const val traktAccessToken = "trakt_access_token"
        const val traktRefreshToken = "trakt_refresh_token"
        const val traktTokenExpiresAt = "trakt_token_expires_at"
        const val traktUsername = "trakt_username"
        const val pixeldrainApiKey = "pixeldrain_api_key"
        const val anilistAccessToken = "anilist_access_token"
    }

    private val prefs: SharedPreferences = run {
        val masterKey = MasterKey.Builder(context.applicationContext)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        try {
            EncryptedSharedPreferences.create(
                context.applicationContext,
                "omniverse_credentials",
                masterKey,
                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
            )
        } catch (_: Throwable) {
            // Fallback to plain prefs if the keystore is unavailable (e.g. corrupted
            // master key on some OEM devices). Better than crashing on launch.
            context.applicationContext.getSharedPreferences("omniverse_credentials_plain", Context.MODE_PRIVATE)
        }
    }

    fun load(): ApiCredentials = ApiCredentials(
        tmdbToken = prefs.getString(K.tmdbToken, "") ?: "",
        tvdbApiKey = prefs.getString(K.tvdbKey, "") ?: "",
        tvdbPin = prefs.getString(K.tvdbPin, "") ?: "",
        traktClientId = prefs.getString(K.traktClientId, "") ?: "",
        traktClientSecret = prefs.getString(K.traktClientSecret, "") ?: "",
        traktAccessToken = prefs.getString(K.traktAccessToken, "") ?: "",
        traktRefreshToken = prefs.getString(K.traktRefreshToken, "") ?: "",
        traktTokenExpiresAt = (prefs.getString(K.traktTokenExpiresAt, "") ?: "").toLongOrNull() ?: 0L,
        traktUsername = prefs.getString(K.traktUsername, "") ?: "",
        pixeldrainApiKey = prefs.getString(K.pixeldrainApiKey, "") ?: "",
        anilistAccessToken = prefs.getString(K.anilistAccessToken, "") ?: "",
    )

    fun save(c: ApiCredentials) {
        prefs.edit().apply {
            putString(K.tmdbToken, c.tmdbToken.trim())
            putString(K.tvdbKey, c.tvdbApiKey.trim())
            putString(K.tvdbPin, c.tvdbPin.trim())
            putString(K.traktClientId, c.traktClientId.trim())
            putString(K.traktClientSecret, c.traktClientSecret.trim())
            putString(K.traktAccessToken, c.traktAccessToken.trim())
            putString(K.traktRefreshToken, c.traktRefreshToken.trim())
            putString(K.traktTokenExpiresAt, c.traktTokenExpiresAt.toString())
            putString(K.traktUsername, c.traktUsername.trim())
            putString(K.pixeldrainApiKey, c.pixeldrainApiKey.trim())
            putString(K.anilistAccessToken, c.anilistAccessToken.trim())
        }.apply()
    }
}

// MARK: - SharedPreferences-backed settings / cache store
// Mirrors the iOS UserSettingsStore (UserDefaults). kotlinx.serialization JSON.

class UserSettingsStore(context: Context) {

    private object K {
        const val settings = "settings"
        const val liveTvSources = "live_tv_sources"
        const val cachedCategories = "cached_categories"
        const val cachedLiveTv = "cached_live_tv"
        const val watchlist = "watchlist"
        const val watchHistory = "watch_history_v1"
        const val lastRefreshedTime = "last_refreshed_time"
    }

    private val prefs: SharedPreferences =
        context.applicationContext.getSharedPreferences("omniverse_settings", Context.MODE_PRIVATE)

    fun loadSettings(): UserSettings = decode(K.settings, UserSettings.serializer()) ?: UserSettings()
    fun saveSettings(s: UserSettings) = encode(K.settings, UserSettings.serializer(), s)

    fun loadLiveTvSources(): List<LiveTvSource> =
        decode(K.liveTvSources, ListSerializer(LiveTvSource.serializer())) ?: emptyList()
    fun saveLiveTvSources(v: List<LiveTvSource>) =
        encode(K.liveTvSources, ListSerializer(LiveTvSource.serializer()), v)

    fun loadCachedCategories(): List<MediaCategory> =
        decode(K.cachedCategories, ListSerializer(MediaCategory.serializer())) ?: emptyList()
    fun saveCachedCategories(v: List<MediaCategory>) =
        encode(K.cachedCategories, ListSerializer(MediaCategory.serializer()), v)

    fun loadCachedLiveTv(): List<LiveTvEntry> =
        decode(K.cachedLiveTv, ListSerializer(LiveTvEntry.serializer())) ?: emptyList()
    fun saveCachedLiveTv(v: List<LiveTvEntry>) =
        encode(K.cachedLiveTv, ListSerializer(LiveTvEntry.serializer()), v)

    fun loadWatchlist(): Set<String> =
        (decode(K.watchlist, ListSerializer(String.serializer())) ?: emptyList()).toSet()
    fun saveWatchlist(v: Set<String>) =
        encode(K.watchlist, ListSerializer(String.serializer()), v.sorted())

    fun loadWatchHistory(): List<WatchProgress> =
        decode(K.watchHistory, ListSerializer(WatchProgress.serializer())) ?: emptyList()
    fun saveWatchHistory(v: List<WatchProgress>) =
        encode(K.watchHistory, ListSerializer(WatchProgress.serializer()), v)

    fun lastRefreshedTime(): Long = prefs.getLong(K.lastRefreshedTime, 0L)
    fun setLastRefreshedTime(ms: Long) = prefs.edit().putLong(K.lastRefreshedTime, ms).apply()

    private fun <T> decode(key: String, serializer: kotlinx.serialization.KSerializer<T>): T? {
        val raw = prefs.getString(key, null) ?: return null
        return try { json.decodeFromString(serializer, raw) } catch (_: Throwable) { null }
    }

    private fun <T> encode(key: String, serializer: kotlinx.serialization.KSerializer<T>, value: T) {
        try {
            prefs.edit().putString(key, json.encodeToString(serializer, value)).apply()
        } catch (_: Throwable) { /* ignore serialization failures */ }
    }
}
